### PR TITLE
Create temporary directory in temporary location

### PR DESCRIPTION
**Pull request by Leonidas-from-XIV**
On my system this leads to `/tmp/dune-install.RANDOM` which is a bit nicer (given it is a `tmpfs`) than just writing a temporary thing into the users home directory.

Of course the downside is that `mv` is not atomic and we might run out of space but these cases seem fairly rare.

### CODE REVIEW

File: static/install [hunk 1]
```diff
@@ -74,7 +74,13 @@ dune_tar_uri="https://get.dune.build/$today/$target/$tar_target"
 
 install_dir="$HOME/.dune"
 bin_dir="$install_dir/bin"
-tmp_dir="$install_dir/tmp"
+
+tmp_dir="$(mktemp -d --tmpdir dune-install.XXXXXXXXXX)"
```
Comment (maiste) [inline static/install:78]: Is `mktemp` available on any POSIX machine or do we need a fallback?

**Review by maiste — APPROVED** [review]
You can merge if the answer to the question below is yes :wink:

File: static/install (hunk 1 above)
Comment (Leonidas-from-XIV) [reply static/install]: I don't think it is POSIX but the command and options seem well supported. I've replaced `--tmpdir` with `-p` to support OpenBSD and it works on Alpine.

[Source](https://stackoverflow.com/a/2792789).

**Comment by maiste** [comment]
Would be nice to have someone on MacOS trying this one.

**Comment by Leonidas-from-XIV** [comment]
@gridbugs Can you give this change a try? I believe it should work on macOS as well according to the info I found online ("very similar to FreeBSD").

**Comment by gridbugs** [comment]
I tested it on macos and I got the error:
```
mktemp: failed to create directory via template 'dune-install.XXXXXXXXXX/tmp.XXXXXXXXXX': No such file or directory
```

But I also get this error on linux. I think it broke when you replaced `--tmpdir` with `-p`. The docs for this argument in the gnu mktemp man page:
```
      -p DIR, --tmpdir[=DIR]
              interpret TEMPLATE relative to DIR; if DIR is not specified, use $TMPDIR if set, else /tmp.  With this option, TEMPLATE must not be an absolute
              name; unlike with -t, TEMPLATE may contain slashes, but mktemp creates only the final component
```

Oddly this implies that the `DIR` value is only optional when using `--tmpdir` but is mandatory when using `-p`, but playing around with mktemp that's consistent with how it behaves.

I think the command should be `mktemp -d -t dune-install.XXXXXXXXXX`. The `-t` causes the base directory to be prepended to the template which results in the directory being placed in "/tmp" or equivalent. Tested with the gnu, macos, and busybox implementations of mktemp.

**Comment by Leonidas-from-XIV** [comment]
Thanks, I've applied your suggestion. The `trap` `rm` started asking questions about deleting a read-only file so I've also forced it to delete the directory. As we create the directory ourselves this shouldn't be a problem, even if it would fail as another run of the script would create a differently named directory.

**Pull request merged**